### PR TITLE
Add section on data manipulation

### PR DESCRIPTION
**Pull request by JonnySpruce**

### CODE REVIEW

File: blog/sudoko_blog_structure.md [hunk 1]
```diff
@@ -4,6 +4,100 @@
...
+
+For example, the below list may be the model's predictions for a particular cell within the puzzle, averaged to total 1 overall:
+
+`[0.01, 0.02, 0.03, 0.01, 0.01, 0.21, 0.03, 0.01, 0.67]`
+
+We can see that most of the predictions are quite small, but the final number in the list is `0.67`, which represents a 67% certainty of it being the number `9`. If the correct answer for that cell was the number `9`, then the ideal answer would have been `[0, 0, 0, 0, 0, 0, 0, 0, 1]`. We can use a loss function to calculate how different the model's prediction was to the correct answer, and then feed that back to the model so it can update its weights to improve future predictions.
+
+#### Changing Scale
+
+So far we have mainly been looking at how one-hot encoding works for a single number, but we need to do this for the whole 9x9 grid, and for all one million puzzles. Thankfully pytorch makes this fairly easy to do the conversion as it provides a function called `one_hot`. Given a tensor, this function automatically converts it into a one-hot encoded version of the tensor, and this-paired with a transform function on the dataset-allows us to convert all of our puzzles.
```
Comment (jstrong-scottlogic) [inline blog/sudoko_blog_structure.md]: ...and this, paired with a transform function on the dataset, allows... - I think such a format would be preferable

File: blog/sudoko_blog_structure.md [hunk 2]
```diff
@@ -4,6 +4,100 @@
...
+
+#### Dataset
+The dataset uses a class provided by pytorch, which we define to have an optional `transform` function. When this function is passed in to the init, it is then applied to the data when accessing a certiain item.
```
Comment (jstrong-scottlogic) [inline blog/sudoko_blog_structure.md]: *certain

File: blog/sudoko_blog_structure.md [hunk 3]
```diff
@@ -4,6 +4,100 @@
 
 We often see neural networks used with sudoku in the realm of computer vision (link to Colin blog) to recognise a sudoku puzzle from a photo or video but rarely do we see machine learning used as a tool to solve sudoku. In this post, we aim to take you through our journey to cracking sudoku with artificial intelligence.
 
+## The Data
+
+A machine learning model is nothing without data to train with, and we initially used the excellent [1 Million Sudoku Puzzles](https://www.kaggle.com/datasets/bryanpark/sudoku) dataset found on Kaggle. It provided each puzzle and solution in a csv format, where the first 81 numbers were the inputs, and the next 81 were the solution. The number 0 was used to represent a blank space in the puzzle. One million puzzles seemed like a good amount to be able to train the model without it being able to easily learn to memorise specific puzzles.
```
Comment (jstrong-scottlogic) [inline blog/sudoko_blog_structure.md:9]: 'without data to train with' reads oddly to me due to the use of both 'without' and 'with.' 'A machine learning model is nothing without data, ... ' may be better.

File: blog/sudoko_blog_structure.md [hunk 4]
```diff
@@ -4,6 +4,100 @@
...
+
+### Data Manipulation
+
+The next step was to consider how we should format the puzzle input. We could just pass the numbers directly, and the model would be able to train to a certain level, however by passing in the numbers as they are creates a relationship between the different values. Machine learning loss functions work by rewarding the model based on how close their prediction is to the actual value, but in sudoku, if the actual answer is `8`, then if the model guesses `7` it is just as incorrect as if it had guessed `1`. This means that we really want the model to be treating each number as a distinct value, and thankfully there is an established way to do this: __one hot encoding__.
```
Comment (jstrong-scottlogic) [inline blog/sudoko_blog_structure.md]: *however passing in

Comment (jstrong-scottlogic) [inline blog/sudoko_blog_structure.md]: Potentially explicitly mention that Sudoku numbers are not ordered and one-hot encoding gets rid of notion of order between them

File: blog/sudoko_blog_structure.md [hunk 5]
```diff
@@ -4,6 +4,100 @@
...
+
+#### One Hot Encoding
+One hot encoding works by converting a distinct set of categorical values into a list of numbers which are either `0` or `1`. The position of the `1` within the array determines which value it represents. For example, if we had three categories, representing `red`, `green`, and `blue`, you could represent them in one-hot encoding using the following values:
```
Comment (jstrong-scottlogic) [inline blog/sudoko_blog_structure.md]: Here we use One hot encoding without a hyphen but later on we use it with, should stick with one choice throughout

File: blog/sudoko_blog_structure.md (hunk 1 above)
Comment (JonnySpruce) [reply blog/sudoko_blog_structure.md]: I've updated it so that it's now two sentences and with commas like you mentioned. Let me know if that looks okay!

**Review by jstrong-scottlogic — APPROVED** [review]
great

**Pull request merged**